### ISSUE TITLE
Godot is cleaning the clipboard(only godot text) on exit

### ISSUE THREAD

**Issue by AlexHolly** [godotengine/godot#4209]
**Operating system or device:**
Ubuntu 14.04 64-bit

Reproduce:
1. open a godot project
2. open a gdscript(or something that has text you can copy from godot)
3. copy(ctrl+c) some text -> paste somewhere outside godot -> works
4. close godot -> paste somewhere outside godot -> empty

**Comment by neikeq** [comment]
I can confirm this on Ubuntu 15.10 64bits. Really annoying :P

**Comment by punto-** [comment]
As far as I remember the clipoard on X is not really a clipboard as in a global place where you store data, it's more like a thing you take ownership of, and then when other applications want to paste, they send the owner a message through X asking for the‎ contents of what you have selected or copied. That's why if the owner is still not running, it doesn't paste anything.. doesn't it work like that for all applications? On 5 April 2016 at 11:00, Ignacio Etcheverry <notifications@github.com> wrote:I can confirm this on Ubuntu 15.10 64bits. Really annoying :P

**Comment by AlexHolly** [comment]
I never noticed this behavior, so let's say this is the case I think the main usage of this is switching between projects and paste something from the previous project.

1.Open Project, copy something
2.Quit to project list(I think here it is also cleaning the clipboard)
3.Open new Project, paste clipboard... it's empty!

**Comment by AlexHolly** [comment]
I never noticed this behavior, so let's say this is the case I think the main usage of this is switching between projects and paste something from the previous project.

1.Open Project, copy something
2.Quit to project list(I think here it is also cleaning the clipboard)
3.Open new Project, paste clipboard... it's empty!

**Comment by neikeq** [comment]
@punto- Most applications I use don't have this problem. There are some exceptions though like Gimp, but they are very rare cases.

**Comment by neikeq** [comment]
From https://wiki.ubuntu.com/ClipboardPersistence

> If you want to fix it as a programmer, you can modify your program to conform to the [ClipboardManager specification](https://freedesktop.org/wiki/ClipboardManager/).

**Comment by AlexHolly** [comment]
When I copy something from the Editor and try to paste it to DBVisualizer I get this error maybe it is related?

```
java.lang.NullPointerException: Failed to retrieve atom name.
	at sun.awt.X11.XlibWrapper.XGetAtomName(Native Method)
	at sun.awt.X11.XAtom.getName(XAtom.java:186)
	at sun.awt.X11.XDataTransferer.getTargetNameForAtom(XDataTransferer.java:165)
	at sun.awt.X11.XDataTransferer.getNativeForFormat(XDataTransferer.java:154)
	at sun.awt.datatransfer.DataTransferer.getFlavorsForFormats(DataTransferer.java:736)
	at sun.awt.datatransfer.ClipboardTransferable.<init>(ClipboardTransferable.java:90)
	at sun.awt.X11.XClipboard.getContents(XClipboard.java:108)
```

**Comment by hpvb** [comment]
After #16156 Godot should behave like other X11 apps and should work about as well as any other program if you don't have a clipboard manager. I don't think this needs a separate fix.

**Comment by hpvb** [comment]
If you don't have a clipboard manager all applications do this though. It's possible that GTK or QT do some magic though. With my latest copy/cut fixes Godot is now compatible with ICCCM clipboard managers, however.

**Comment by AlexHolly** [comment]
I installed `CopyQ` for testing works fine.


parcellite did not work.
> From https://wiki.ubuntu.com/ClipboardPersistence

Would be nice to have a solution without a clipboard manager  :)

**Comment by hpvb** [comment]
That is odd, I use parcellite myself and Godot seems to play nice with it after #16156. What problem are you still seeing?

**Comment by hpvb** [comment]
Just as a clarification as to how X11 clipboards work (Godot now implements this correctly)

1) Godot -> I'd like ownership of the clipboard -> X server
2) Gedit -> I'd like to paste some date off of the clipboard, what formats are supported on there now? -> X server
3) X server -> Tell me what formats of data you have on the clipboard -> Godot
4) Godot -> Well kind sir, I have UTF-8 data and ASCII text -> X server
5) Xserver -> There is UTF-8 data and ASCII text on the clipboard -> Gedit
6) Gedit -> I'd love myself some UTF-8 data please -> Xserver
7) Xserver -> Please send me UTF-8 data from your clipboard -> Godot
8) Godot -> You still here? Fine Here's some data -> Xserver
9) Xserver -> No problemo, here's some data -> Gedit

Now if Godot goes away it can't be asked for any of that anymore. And the data is gone, it never gets copied it just gets streamed between the cutter and paster *when it is required* no sooner.

What a clipboard manager does is basically constantly running the following loop
10 Do I own the clipboard? if so goto 10
20 Get data from the clipboard in UTF-8 or Text and store it
30 Take ownership of the clipboard
40 goto 10

In this case if the application goes away the clipboard manager still has a copy.

**Comment by hpvb** [comment]
OK, while my previous explanation was correct for older applications (and this *is* what godot does now) I was completely unaware of SAVE_TARGETS... I'll implement this. Sorry @AlexHolly you were entirely correct. This can be fixed by Godot and it will.

**Comment by hpvb** [comment]
Hmm, I've implemented SAVE_TARGETS for Godot but it doesn't seem to work for me. I tried other (GTK) apps and at least on gnome3 if I stop parcellite I don't have saved clipboards after exit. I need to investigate a bit further. It's possible that gnome-shell won't take over duties as the clipboard manager once another has been started once in the session.

**Comment by hpvb** [comment]
This will be a separate PR btw, #16156 will continue to be limited to correcting interaction with normal clipboard managers, and fixes some X protocol errors.

A full SAVE_TARGETS implementation will require some refactoring of the whole way xevents are handled. This is not a problem but we need to figure out how best to do this, and I need to find out how it's supposed to work under modern desktops so I can meet the normal users' expectations with our implementation.

**Comment by Anutrix** [comment]
Update needed.
